File: frontend/src/types/payment.ts
export type PaymentStatus =
  | 'SUCCESS'
  | 'FAILED'
  | 'PENDING'
  | 'ABANDONED'
  | 'EXPIRED'
  | 'RECOVERED';

export type FailureReason =
  | 'INSUFFICIENT_FUNDS'
  | 'BANK_TIMEOUT'
  | 'CARD_DECLINED'
  | 'NETWORK_ERROR'
  | 'AUTHENTICATION_FAILED'
  | 'LIMIT_EXCEEDED'
  | 'UNKNOWN';

export interface Customer {
  id: number;
  customer_id: string;
  name: string;
  email: string;
  phone?: string | null;
  account_age_days: number;
  total_payments: number;
  successful_payments: number;
  failed_payments: number;
  average_payment_amount: number;
  created_at: string;
  updated_at: string;
}

export interface Merchant {
  id: number;
  merchant_id: string;
  name: string;
  business_category: string;
  api_key_prefix: string;
  created_at: string;
  updated_at: string;
}

export interface Payment {
  id: number;
  payment_id: string;
  customer_id: number;
  merchant_id: number;
  amount: number | string;
  currency: string;
  payment_method: string;
  status: PaymentStatus;
  failure_reason?: FailureReason | string | null;
  retry_count: number;
  customer?: Customer | null;
  merchant?: Merchant | null;
  created_at: string;
  updated_at: string;
}

export interface PaymentFilterParams {
  skip?: number;
  limit?: number;
  status?: string;
  failure_reason?: string;
  search?: string;
}

File: backend/app/schemas/payment.py
from datetime import datetime
from decimal import Decimal
from typing import Optional
from pydantic import BaseModel, ConfigDict
from app.models.enums import PaymentStatus, FailureReason
from app.schemas.customer import CustomerRead
from app.schemas.merchant import MerchantRead


class PaymentBase(BaseModel):
    payment_id: str
    customer_id: int
    merchant_id: int
    amount: Decimal
    currency: str = "INR"
    payment_method: str
    status: PaymentStatus = PaymentStatus.PENDING
    failure_reason: Optional[FailureReason] = None
    retry_count: int = 0


class PaymentCreate(PaymentBase):
    pass


class PaymentRead(PaymentBase):
    id: int
    customer: Optional[CustomerRead] = None
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


class PaymentDetail(PaymentRead):
    customer: Optional[CustomerRead] = None
    merchant: Optional[MerchantRead] = None

    model_config = ConfigDict(from_attributes=True)
